### ISSUE TITLE
feat: Supervised Pipeline Learning Service (DeBERTa Fine-Tuning)

### ISSUE THREAD

**Issue by epappas** [epappas/llmtrace#44]
## Executive Summary

A standalone Python microservice that fine-tunes DeBERTa-v3-base-prompt-injection-v2 on user-specific traffic using LLM-as-a-Judge decisions as ground truth. Extracts training data from ClickHouse/PostgreSQL, trains with configurable hyperparameters and class balancing, exports in safetensors format (the Rust proxy uses Candle with safetensors, NOT ONNX), and publishes versioned model artifacts. Retraining is manually triggered; the proxy requires restart to load the new model.

## Problem Statement

The base DeBERTa model is trained on a general-purpose corpus and may not capture domain-specific patterns in a given organization's traffic. False positives on legitimate business terminology and false negatives on novel attack vectors erode trust. There is currently no feedback loop: ensemble decisions are stored but never used to improve the model. This pipeline closes the loop using curated Judge labels as ground truth, producing a model calibrated to actual traffic distribution.

## Important Technical Constraint

The Rust proxy uses **Candle with safetensors**, NOT ONNX Runtime. The fine-tuning service must export: `model.safetensors`, `config.json`, `tokenizer.json`. PyTorch's `safetensors` library writes this format natively. No ONNX conversion needed.

## System Context

```
+-------------------+    Extract    +----------------------------+
| ClickHouse        | <----------> | Pipeline Learning Service  |
| (spans: prompt,   |              | (Python)                   |
|  security_score,  |              |                            |
|  security_findings)|    Export    |  - Dataset extraction      |
+-------------------+  ----------> |  - Quality validation      |
                                   |  - Fine-tune DeBERTa       |
+-------------------+              |  - Export safetensors       |
| PostgreSQL        |              |  - Model registry          |
| (judge_verdicts)  | <--extract-- +------------+---------------+
+-------------------+                           |
                                    Publish artifact
                                                |
                                                v
                                   +----------------------------+
                                   | Model Registry (local/S3)  |
                                   +------------+---------------+
                                                |
                                    Manual copy / volume mount
                                                |
                                                v
                                   +----------------------------+
                                   | LLMTrace Proxy (Rust)      |
                                   | loads model.safetensors    |
                                   | via ml_cache_dir           |
                                   +----------------------------+
```

### Python Project Structure

```
services/pipeline-learning/
  pyproject.toml
  Dockerfile
  config.example.yaml
  src/
    pipeline_learning/
      __init__.py
      cli.py                    # click-based CLI
      api.py                    # FastAPI for trigger/status
      config.py                 # Pydantic config model
      extract/
        clickhouse.py           # Extract traces + findings
        postgres.py             # Extract judge decisions
        merger.py               # Join traces with labels
      dataset/
        schema.py               # Dataset record types
        validator.py            # Quality checks
        splitter.py             # Train/val/test split
        balancer.py             # Class imbalance handling
      training/
        trainer.py              # HuggingFace Trainer wrapper
        metrics.py              # F1, precision, recall
        callbacks.py            # Logging, checkpointing
      export/
        safetensors_export.py   # Model + tokenizer + config export
        registry.py             # Version management
      quality_gate/
        evaluator.py            # New vs baseline comparison
        report.py               # Evaluation report
  tests/
    conftest.py
    test_extract.py
    test_dataset.py
    test_training.py
    test_export.py
    test_quality_gate.py
```

### Key Dependencies

```toml
[project]
name = "llmtrace-pipeline-learning"
requires-python = ">=3.11"
dependencies = [
    "transformers>=4.44", "torch>=2.2", "datasets>=3.0",
    "safetensors>=0.4", "peft>=0.12", "evaluate>=0.4",
    "scikit-learn>=1.4", "clickhouse-connect>=0.7",
    "psycopg[binary]>=3.1", "fastapi>=0.115", "uvicorn>=0.30",
    "pydantic>=2.7", "pydantic-settings>=2.3", "click>=8.1",
    "structlog>=24.1", "boto3>=1.34",
]
```

### Dataset Schema

```json
{
  "text": "Ignore all previous instructions...",
  "label": 1,
  "judge_confidence": 0.95,
  "judge_verdict": "injection",
  "judge_reasoning": "Direct instruction override attempt",
  "original_ensemble_score": 85,
  "original_ensemble_label": "prompt_injection",
  "trace_id": "550e8400-...",
  "tenant_id": "123e4567-...",
  "timestamp": "2026-03-01T14:30:00Z",
  "token_count": 12
}
```

### FR-01: Training Data Extraction
- Query ClickHouse spans for prompt, security_score, security_findings
- Query PostgreSQL for judge decisions (verdict, reasoning, confidence)
- Join on trace_id; supports time-range and tenant filtering
- Deduplicates identical prompt texts; skips unlabeled records
- **AC:** Produces standardized dataset with all fields from schema

### FR-02: Dataset Quality Validation
- Reject datasets < 500 total samples or < 50 per class
- Report class distribution, label agreement rate, text length stats
- Flag low-confidence labels (below configurable threshold, default 0.7)
- Detect duplicates/near-duplicates
- **AC:** Validation report as JSON; sub-threshold datasets rejected with clear error

### FR-03: Class Imbalance Handling
- Three strategies: `weighted_loss` (default), `oversample`, `undersample`
- Log chosen strategy and resulting distribution
- **AC:** Each strategy produces correctly balanced training set

### FR-04: Train/Validation/Test Split
- Default: 70/15/15 stratified split
- Optional temporal split (train on older, test on newer)
- No data leakage (identical texts go to same split)
- **AC:** Split statistics in training report; class ratios preserved

### FR-05: Fine-Tuning Pipeline
- Base model: `protectai/deberta-v3-base-prompt-injection-v2`
- HuggingFace Trainer with: lr=2e-5, batch=16, epochs=3, weight_decay=0.01
- LoRA support via PEFT (r=16, alpha=32, targets=query_proj+value_proj) -- 8GB VRAM
- Full fine-tuning also available -- 16GB VRAM
- Early stopping (patience=2 on val F1), evaluation every epoch
- **AC:** Training completes, checkpoints saved, metrics logged per epoch

### FR-06: Model Export in Candle-Compatible Format
- Export: `model.safetensors`, `config.json`, `tokenizer.json`
- LoRA models: merge weights before export
- Validate exported model matches PyTorch predictions (max deviation < 1e-5)
- Create `model_card.json` with metadata
- **AC:** Exported model loadable by Candle with identical predictions

### FR-07: Model Versioning and Registry
- Version format: `v{YYYYMMDD}-{HHmmss}-{short_sha}`
- Artifact directory: model files + evaluation report + training config
- Registry manifest (`registry.json`): status = candidate | approved | active | retired
- Supports local filesystem (default) and S3
- CLI: `list-models`, `promote-model`, `retire-model`
- **AC:** Multiple versions coexist; only one `active` at a time

### FR-08: Quality Gate Evaluation
- Compare fine-tuned vs baseline on held-out test set
- Metrics: accuracy, F1, precision, recall, AUC-ROC
- Gate passes if: F1 >= baseline - 0.02, recall >= baseline - 0.05, precision >= baseline
- Failed gate prevents promotion to `approved`
- **AC:** Side-by-side report generated; regression blocks promotion

### FR-09: API and CLI Interface
- CLI commands:
  - `pipeline-learning extract --since 30d --output dataset.jsonl`
  - `pipeline-learning validate --dataset dataset.jsonl`
  - `pipeline-learning train --config training.yaml`
  - `pipeline-learning evaluate --model-version <v> --test-set test.jsonl`
  - `pipeline-learning list-models`
  - `pipeline-learning promote --version <v>`
- FastAPI endpoints (optional):
  - `POST /api/v1/train` -> returns run_id
  - `GET /api/v1/runs/{run_id}` -> status + metrics
  - `GET /api/v1/models` -> list versions
  - `POST /api/v1/models/{version}/promote` -> promote
- **AC:** CLI and API produce identical results

### FR-10: Deployment Integration
- `promote` copies artifact to `ml_cache_dir` path
- Directory structure matches what hf-hub produces (proxy loads transparently)
- Proxy restart required to load new model
- Supports volume mounting in Docker/K8s
- **AC:** Promoted model loadable by proxy after restart

## Configuration

```yaml
sources:
  clickhouse:
    url: "http://localhost:8123"
    database: "llmtrace"
  postgres:
    url: "postgres://llmtrace:llmtrace@localhost:5432/llmtrace"

extraction:
  since_days: 30
  tenant_id: null
  min_judge_confidence: 0.7
  deduplicate: true

validation:
  min_total_samples: 500
  min_class_samples: 50
  max_duplicate_ratio: 0.3

splitting:
  strategy: "stratified"    # "stratified" | "temporal"
  train_ratio: 0.70
  val_ratio: 0.15
  test_ratio: 0.15

imbalance:
  strategy: "weighted_loss"  # "weighted_loss" | "oversample" | "undersample"

training:
  base_model: "protectai/deberta-v3-base-prompt-injection-v2"
  use_lora: true
  lora_r: 16
  lora_alpha: 32
  learning_rate: 2.0e-5
  batch_size: 16
  epochs: 3
  weight_decay: 0.01
  warmup_ratio: 0.1
  max_seq_length: 512
  early_stopping_patience: 2
  device: "auto"
  mixed_precision: "fp16"

quality_gate:
  max_f1_regression: 0.02
  max_recall_regression: 0.05
  require_precision_improvement: true

registry:
  backend: "local"
  local_path: "./models"
  s3_bucket: ""
  s3_prefix: "llmtrace-models/"

proxy:
  ml_cache_dir: "/home/llmtrace/.cache/llmtrace/models"
```

### Training (per run)
- `training_loss`, `validation_loss` per epoch
- `validation_f1`, `validation_precision`, `validation_recall` per epoch
- `test_f1`, `test_auc_roc` (final)
- `training_duration_seconds`, `gpu_memory_peak_mb`
- `dataset_size_total`, `dataset_positive_ratio`

### Operational (Prometheus, if API mode)
- `pipeline_training_runs_total` (by status)
- `pipeline_training_duration_seconds`
- `pipeline_quality_gate_passed_total`

## Validation Plan

1. **Unit tests:** Extraction SQL, validation logic, split logic, class balancing
2. **Integration tests:** Against local ClickHouse/PostgreSQL with fixtures
3. **Training smoke test:** 1-epoch on 100 samples, verify safetensors export loadable
4. **Candle compatibility:** Load exported model in Rust test program, verify forward pass matches PyTorch
5. **Quality gate test:** Verify logic with mocked baseline/candidate metrics
6. **E2E:** Extract from populated DB -> train -> export -> load in proxy -> inference on known samples

## Implementation Sequence

| Phase | Scope | Timeline |
|-------|-------|----------|
| 1 | Data pipeline: extraction, merger, validation, splitting, CLI | Week 1-2 |
| 2 | Training: Trainer wrapper, balancing, metrics, callbacks, CLI | Week 3-4 |
| 3 | Export and registry: safetensors export, versioning, quality gate | Week 5 |
| 4 | Integration and API: FastAPI, deployment integration, Dockerfile, E2E | Week 6 |

## Known Limitations

1. **Batch-only**: No online/continuous learning; manually triggered
2. **Proxy restart required**: No hot-reload of models at runtime
3. **No federated training**: One model; per-tenant requires separate runs + instances
4. **Judge quality dependency**: Model only as good as Judge labels
5. **LoRA merge required**: Candle doesn't support LoRA adapters natively; merge before export
6. **Safetensors, not ONNX**: Proxy uses Candle; ONNX is unnecessary

## Risk Assessment

| Risk | Impact | Likelihood | Mitigation |
|------|--------|------------|------------|
| Insufficient labeled data | High | High | Min thresholds (FR-02); recommend 30+ days of Judge operation |
| Fine-tuned model regresses | High | Medium | Quality gate (FR-08) blocks promotion on regression |
| GPU unavailability | Medium | Medium | LoRA for 8GB VRAM; CPU-only supported |
| Weight format incompatibility with Candle | Critical | Low | Automated compatibility test in CI |
| Class imbalance distorts training | Medium | High | Three balancing strategies; default weighted_loss |
| Judge label noise | High | Medium | Filter low-confidence labels; report agreement metrics |

## Dependencies

- **Depends on:** #43 for training labels in `judge_verdicts` table
- **Integration with:** LLMTrace Proxy (model loading via `ml_cache_dir`)
